Track Covariance Map

ax3l opened the pull request:
Add a new runtime mode, tracking of 2nd moments in a covariance matrix.

- [x] generalize initialization #815
- [x] add map initialization to each element
- [x] add maps for core elements (drift, quad, bend, shortrf)
- [x] push loop
- [x] diagnostics
- [x] example(s): input file
- [x] example(s): Python
- [x] example(s): analysis / NaN pandas issue
- [x] documentation

Follow-up:  add the implementation of space charge kicks (in envelope mode), following Issue #826

ax3l committed:
[Draft] `track_covariance_map`

ax3l commented on src/ImpactX.cpp:
```diff
@@ -336,4 +336,35 @@ namespace impactx {
+                }
+
+                // push Covariance Matrix
+                cm = cm * element.transport_map(ref);
```
Correct would be: `R_T * cm * R`

cemitch99 replied:
Transpose is on the right:  `R * cm * R_T`

ax3l replied:
Thanks! Dang, we had a 50% chance to get it right by guessing :D

ax3l replied:
fixed

cemitch99 committed:
Add Quad R-map.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Update Quad.H

cemitch99 committed:
Add Sbend R-map.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Add shortrf R-map.

ax3l committed:
Spaces

cemitch99 commented on src/particles/elements/mixin/lineartransport.H:
```diff
@@ -24,27 +24,35 @@ namespace impactx::elements::mixin
+            // small trick to force every derived class has to implement a method transport_map
+            // (w/o using a purely virtual function)
+            T_Element& element = *static_cast<T_Element*>(this);
+            cm *= element.transport_map(cm);
```
Is this logic right?  The function `transport_map` in each element takes the reference particle data as input, but this appears to call `transport_map` with the covariance matrix `cm` as input.  Also, if the intent is to push the covariance matrix by the matrix `R`, then the math should look like `cm = R * cm * R^T`, which appears also `ImpactX.cpp`.

ax3l replied:
Oh, that is outdated. I only fixed it in `ImpactX.cpp`.

cemitch99 commented:
Regarding user control of the run mode, I propose `algo.envelope_mode = true/false`, with false as the default (particle tracking).

cemitch99 committed:
Add note on diagnostics in ImpactX.cpp

ax3l commented:
We could also make it more general: `algo.method = `
- `envelope`
- `reference_particle`
- `tracking` (default)

That way we have one flag for all planned modes and we can continue to add modes beyond 3 in the future :D (e.g., ensemble/batch variants of these modes or so)

cemitch99 commented:
What about `algo.mode = track_envelope, track_ref_part, track_particles`, or similar?  This could reflect the name of the function used in the push loop.  I think "method" is maybe not descriptive enough, and could apply to any algorithm.

ax3l commented:
Perfect

ax3l commented:
Oh, or instead of mode we name it `algo.track = envelope, ref_part, particles`.

cemitch99 commented:
That works, too.  FYI:  I used `algo.mode = track_envelopes` in the diagnostics PR, but it's a trivial change.

ax3l committed:
Fix: LinearTransport Mixin

ax3l committed:
Elements: Use LinearTransport Mixin

ax3l committed:
Linear Transport: Python Push

ax3l committed:
Elements: Throw Not Implemented Error

For elements that do not yet have maps.

cemitch99 committed:
Add covariance map diagnostics (#4)

* Update DiagnosticOutput & Reduced Beam Diagnostics

---------

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

ax3l commented:
I will fix the Doxygen warning/error later, please ignore for now.

ax3l committed:
Diagnostics in Envelope Loop

ax3l committed:
Merge remote-tracking branch 'mainline/development' into topic-track-covariance-map

ax3l committed:
Finish Restructuring of ASCII Diags

ax3l committed:
Python: Simplify Map6x6 Exposure

ax3l committed:
Add `algo.track` Control Logic

ax3l committed:
Fix Doxygen

ax3l committed:
CI: Doxygen Ubuntu 22.04 -> 24.04

ax3l committed:
Doc: New User Introduction

ax3l commented:
@cemitch99 I started a new, general user introduction:
https://impactx--814.org.readthedocs.build/en/814/usage/how_to_run.html

ax3l committed:
Doc: APIs

ax3l commented:
I also documented the new APIs for the new run modes. Now we only need to add an example.

ax3l committed:
RBC: Keep Same Columns

ax3l commented on src/particles/diagnostics/ReducedBeamCharacteristics.cpp:
```diff
@@ -403,7 +406,183 @@ namespace impactx::diagnostics
+        auto const nan = std::numeric_limits<amrex::ParticleReal>::quiet_NaN();
+
+        std::unordered_map<std::string, amrex::ParticleReal> data;
+        data["x_mean"] = nan;
```
We can check if most of those are truly undefined or instead are just "trivial" values (like `0`). I would plot the trivial ones as well, because it makes the columns available more consistent and one can reuse plot scripts for envelope and tracking sims that way.

ax3l commented on src/particles/diagnostics/DiagnosticOutput.cpp:
```diff
@@ -9,226 +9,221 @@
+    prepare_header (
+        amrex::AllPrintToFile & file_handler,
+        OutputType otype,
+        bool has_charge,
```
Bug: remove `has_charge` and write all columns unconditionally.

cemitch99 committed:
Modify ImpactX.cpp to fix envelope tracking loop. (#5)

ax3l committed:
Fix Doc Table Formatting

cemitch99 committed:
Add DipEdge support & IOTA example (#6)

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 commented:
Note:  The NaNs appearing in `reduced_beam_characteristics.*` appear to cause a problem when using the strategy from PR #822 to write the CI analysis scripts.  Need to find a workaround or replace the NaNs.

cemitch99 commented on docs/source/usage/how_to_run.rst:
```diff
@@ -3,87 +3,89 @@
-#. make sure the ImpactX **executable** is either copied into this directory or in your ``PATH`` `environment variable <https://en.wikipedia.org/wiki/PATH_(variable)>`__
-#. add an **inputs file** and on :ref:`HPC systems <install-hpc>` a **submission script** to the directory
-#. run
+When to run: Physics Models
```
It's great to have this summary here, but I don't think the "when", "what", "how", headings are very helpful.  For example, when I read "When to run", my first response is:  "When I need to model something."  I suggest more direct, uniformly-phrased questions like:  "How to select a run mode", "How to select a user interface", "How to select a computing system", and "How to run the code".  I'll add some possible suggestions.

ax3l replied:
Thanks! Yes, I am very flexible on these headings, thanks for the suggestions!

cemitch99 commented on docs/source/usage/how_to_run.rst:
```diff
@@ -3,87 +3,89 @@
-#. make sure the ImpactX **executable** is either copied into this directory or in your ``PATH`` `environment variable <https://en.wikipedia.org/wiki/PATH_(variable)>`__
-#. add an **inputs file** and on :ref:`HPC systems <install-hpc>` a **submission script** to the directory
-#. run
+When to run: Physics Models
```
```suggestion
How to select a tracking mode
```

cemitch99 commented on docs/source/usage/how_to_run.rst:
```diff
@@ -3,87 +3,89 @@
 
-1. Run Directory
-----------------
+ImpactX most powerful and central physics model is its symplectic particle tracker with collective effects (space charge, CSR, wakefields, etc.).
```
```suggestion
ImpactX can be run using any of three distinct tracking modes.  ImpactX's most powerful tracking mode makes use of symplectic particle tracking with collective effects included (space charge, CSR, wakefields, etc.).
```

cemitch99 replied:
I would not call these "physics models".  They really are modes that differ due to what is being tracked--for example, for linear beamline elements (default) the physical model is the same in all three cases.

ax3l replied:
Agreed, thx!

cemitch99 commented on docs/source/usage/how_to_run.rst:
```diff
@@ -3,87 +3,89 @@
-1. Run Directory
-----------------
+ImpactX most powerful and central physics model is its symplectic particle tracker with collective effects (space charge, CSR, wakefields, etc.).
+Additionally, ImpactX provides two simplified physics models to aid scientists through every step, from beamline inception to operation:
```
```suggestion
Additionally, ImpactX provides two simplified tracking modes to aid scientists through every step, from beamline inception to operation:
```

cemitch99 commented on docs/source/usage/how_to_run.rst:
```diff
@@ -3,87 +3,89 @@
-----------------
+ImpactX most powerful and central physics model is its symplectic particle tracker with collective effects (space charge, CSR, wakefields, etc.).
+Additionally, ImpactX provides two simplified physics models to aid scientists through every step, from beamline inception to operation:
+only tracking of the reference particle orbit, or tracking of the beam envelope (covariance matrix) through linearized transport maps.
```
```suggestion
only tracking of the reference particle orbit, or tracking of the beam envelope (6x6 covariance matrix) through linearized transport maps.
```

cemitch99 commented on docs/source/usage/how_to_run.rst:
```diff
@@ -3,87 +3,89 @@
-.. code-block:: bash
 
-   mkdir -p <run_directory>
+What to run: User Interfaces
```
```suggestion
How to select a user interface
```

cemitch99 commented on docs/source/usage/how_to_run.rst:
```diff
@@ -3,87 +3,89 @@
-   impactx<TAB>
 
-Hitting the ``<TAB>`` key will suggest available ImpactX executables as found in your ``PATH`` `environment variable <https://en.wikipedia.org/wiki/PATH_(variable)>`__.
+How to run
```
I propose moving the "How to run" section to the end of this page, and specifying that the steps described are given for a local computer.  For details on how to proceed on HPC machines, we can add a link to the existing documentation.

ax3l replied:
Agreed. Let us quickly connect if you would do another change here to discuss.

ax3l replied:
Resolved offline: ok for now, we will do a theory section update as a follow-up and will refine docs further.

cemitch99 commented on docs/source/usage/how_to_run.rst:
```diff
@@ -3,87 +3,89 @@
-   mpirun -np <n_ranks> ./impactx <input_file>
 
-or
+Where to run
```
```suggestion
How to select a computing system
```

cemitch99 commented on docs/source/usage/parameters.rst:
```diff
@@ -11,6 +11,20 @@ This documents how to use ImpactX with an input file (``impactx input_file.in``)
 
+.. _running-cpp-parameters-mode:
+
+Physics Models
```
```suggestion
Tracking Modes
```

ax3l replied:
Fun! I actually called them `modes`  in the internal anchor (2 lines above) and then used for some reason `Model` in the visible text xD

cemitch99 commented on docs/source/usage/parameters.rst:
```diff
@@ -11,6 +11,20 @@ This documents how to use ImpactX with an input file (``impactx input_file.in``)
+
+* ``algo.track`` (``string``)
+
+  Physical tracking model to run:
```
```suggestion
  Mode that specifies how the beam is tracked:
```

cemitch99 committed:
Add example analysis scripts (#7)

* Add example analysis scripts and testing.

* Update analysis_fodo_envelope.py

* Update analysis_iotalattice_envelope.py

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l committed:
Documentation Improvements

Co-authored-by: Chad Mitchell <46825199+cemitch99@users.noreply.github.com>

ax3l commented on examples/fodo/README.rst:
```diff
@@ -76,3 +76,65 @@ You can run the following script to visualize the beam evolution over time:
+due to$
+
+In this test, the initial and final values of :math:`\lambda_x`, :math:`\lambda_y`, :math:`\lambda_t`, :math:`\epsilon_x`, :math:`\epsilon_y`, and :math:`\epsilon_t` must agree with nominal
+valu$
```
Copy paste from a cropped terminal :)

ax3l committed:
Doc Formatting

ax3l committed:
Fix Analysis Script Permissions

`chmod a+rx *py`

ax3l committed:
Remove `has_charge`

Temporary/outdated from earlier code draft. Fix column headers.

ax3l committed:
CMake: Clean Up FODO Example Names

Unify spellings

ax3l committed:
Example: Remove Space Charge Properties

ax3l committed:
Reorder `amr_data` for modes

Avoid confusing/mixing data.

ax3l committed:
Python Implementation

ax3l committed:
FODO Envelope Python Example

ax3l committed:
IOTA Example: Remove `sigt` Analysis

ax3l committed:
IOTA Lattice Envelope: Python Example

ax3l committed:
Python: Make Interactive Use Safer

People might forget to call `sim.init_envelope`, in which
case we do not want them to calculate with zero-valued ref or cm.

ax3l committed:
Complete `AmrCoreData::track_particles`

Move the fields over as well, only used in particle tracking.

cemitch99 committed:
Add comparison of sig_tf to analysis.

ax3l committed:
Cleanup: `PushAll`

This continues to only push particles.

cemitch99 committed:
Add sigt cf for iota analysis. (#8)

ax3l commented on src/python/elements.cpp:
```diff
@@ -72,6 +72,22 @@ namespace
+        using Element = typename T_PyClass::type;  // py::class<T, options...>
+
+        cl.def("push",
+               [](Element & el, Map6x6 & cm, RefPart const & AMREX_RESTRICT ref) {
```
```suggestion
               [](Element & el, Map6x6 & cm, RefPart const & ref) {
```

ax3l committed:
Simplify register_envelope_push

ax3l commented on src/particles/diagnostics/DiagnosticOutput.cpp:
```diff
@@ -9,226 +9,216 @@
+            if (compute_eigenemittances)
+            {
+                file_handler << " "
+                             << "emittance_xn" << " " << "emittance_yn" << " " << "emittance_tn";
```
Copy-paste bug: Fix in #864
